packages/client/hmi-client/src/services/models/simulation-service.ts

```diff
@@ -195,3 +197,134 @@ export async function makeEnsembleCiemssCalibration(params: EnsembleCalibrationC
 		return null;
 	}
 }
+
+// add a simulation in progress if it does not exist
+const addSimulationInProgress = (node: WorkflowNode, runIds: string | string[]) => {
+	const state = cloneDeep(node.state);
+
+	if (!state.simulationsInProgress) {
+		state.simulationsInProgress = [];
+	}
+
+	if (typeof runIds === 'string') {
+		// single run id
+		if (!state.simulationsInProgress.find((simulation) => simulation === runIds)) {
+			state.simulationsInProgress.push(runIds);
+			return state;
+		}
+		return null;
+	}
+
+	// array of run ids
+	let allExist = false;
+	runIds.forEach((runId) => {
+		if (!state.simulationsInProgress.includes(runId)) {
+			state.simulationsInProgress.push(runId);
+			allExist = true;
+		}
+	});
+	return allExist ? state : null;
+};
+
+// delete a simulation in progress
+const deleteSimulationInProgress = (node: WorkflowNode, runIds: string | string[]) => {
+	const state = cloneDeep(node.state);
+
+	if (state.simulationsInProgress) {
+		// handle single run id
+		if (typeof runIds === 'string') {
+			state.simulationsInProgress = state.simulationsInProgress.filter(
+				(simulation) => simulation !== runIds
+			);
+			return state;
+		}
+
+		// handle array of run ids
+		runIds.forEach((runId) => {
+			const index = state.simulationsInProgress.indexOf(runId);
+			if (index !== -1) {
+				state.simulationsInProgress.splice(index, 1);
+			}
+		});
+		return state;
+	}
+
+	return null;
+};
+
+const querySimulationInProgress = (node: WorkflowNode): string[] => {
+	const state = node.state;
+	if (state.simulationsInProgress && state.simulationsInProgress.length > 0) {
+		return state.simulationsInProgress; // getStatus(state.simulationsInProgress[0])
+	}
+
+	return [];
+};
+
+// handle all simulation in progress operations
+export function handleSimulationsInProgress(
+	operation: SimulationStateOperation,
+	node: WorkflowNode,
+	runId: string | string[] = ''
+) {
+	switch (operation) {
+		case SimulationStateOperation.ADD:
+			return addSimulationInProgress(node, runId);
+		case SimulationStateOperation.DELETE:
+			return deleteSimulationInProgress(node, runId);
+		case SimulationStateOperation.QUERY:
+			return querySimulationInProgress(node);
+		default:
+			break;
+	}
+	return null;
+}
```

Comment:
What is the purpose of this function? It seems like it is just a pass-thru facade, why wouldn't we just directly call each `addSimulationInProgress/deleteSimulationInProgress/querySimulationInProgress` directly given that you have to specify the operation anyway?


Also we wouldn't generally want to return `null` as the new state, because it will mess up other things. We didn't do this part particularly well due to lack of time, but each "operation" does have a statically defined state definition. of what it would expect. For example, see

workflow/calibrate-operation-julia.ts#CalibrationOperationStateJulia and the initState function

Reply:
@mwdchang You're right, I've got rid of this pass-thru function entirely it wasn't necessary.  As for the add/delete query they are no longer returning null but always return a state.  When this state is return it is checked against the older state to see if we should be emitting a state change or not.

I have also added the `simulationsInProgress` key to all the node states interfaces.